README.md
# Python Library for quering OSS Index

![GitHub Workflow Status](https://img.shields.io/github/workflow/status/sonatype-nexus-community/ossindex-python/Python%20CI)
![Python Version Support](https://img.shields.io/badge/python-3.6+-blue)
![PyPI Version](https://img.shields.io/pypi/v/ossindex-lib?label=PyPI&logo=pypi)
[![Documentation](https://readthedocs.org/projects/ossindex-library/badge/?version=latest)](https://readthedocs.org/projects/ossindex-library)
[![GitHub license](https://img.shields.io/github/license/sonatype-nexus-community/ossindex-python)](https://github.com/sonatype-nexus-community/ossindex-python/blob/main/LICENSE)
[![GitHub issues](https://img.shields.io/github/issues/sonatype-nexus-community/ossindex-python)](https://github.com/sonatype-nexus-community/ossindex-python/issues)
[![GitHub forks](https://img.shields.io/github/forks/sonatype-nexus-community/ossindex-python)](https://github.com/sonatype-nexus-community/ossindex-python/network)
[![GitHub stars](https://img.shields.io/github/stars/sonatype-nexus-community/ossindex-python)](https://github.com/sonatype-nexus-community/ossindex-python/stargazers)

----

This OSSIndex module for Python provides a common interface to querying the [OSS Index](https://ossindex.sonatype.org/).

This module is not designed for standalone use. If you're looking for a tool that can detect your application's dependencies
and assess them for vulnerabilities against the OSS Index, perhaps you should check out 
[Jake](https://github.com/sonatype-nexus-community/jake).

You can of course use this library in your own applications.

## Installation

Install from pypi.org as you would any other Python module:

```
pip install ossindex-lib
```

## Usage

First create an instance of `OssIndex`, optionally enabling local caching
```
o = OssIndex()
```

Then supply a `List` of [PackageURL](https://github.com/package-url/packageurl-python) objects that you want to ask
OSS Index about. If you don't want to care about generating this list yourself, perhaps look to a tool like [Jake](https://github.com/sonatype-nexus-community/jake)
(which uses this library) and will do all the hard work for you!

As a quick test, you could run:
```
o = OssIndex()
results: List[OssIndexComponent] = o.get_component_report(packages=[
    PackageURL.from_string(purl='pkg:pypi/pip@19.2.0')
])
for r in results:
    print("{}: {} known vulnerabilities".format(r.get_coordinates(), len(r.get_vulnerabilities())))
    v: Vulnerability
    for v in r.get_vulnerabilities():
        print('    - {}'.format(str(v)))
```

... which would output something like ...
```
pkg:pypi/pip@19.2.0: 1 known vulnerabilities
    - <Vulnerability id=e4c955a3-2004-472e-920b-783fea46c3cd, name=OSSINDEX-783f-ea46-c3cd, cvss_score=3.6>
```

## Logging

This library send log events to a standard Python `logger` named `ossindex`. You can configure the logger to output as
required through the standard [Python logging configuration](https://docs.python.org/3/library/logging.config.html).

## Todos

1. Support authentication against OSS Index

## Python Support

We endeavour to support all functionality for all [current actively supported Python versions](https://www.python.org/downloads/).
However, some features may not be possible/present in older Python versions due to their lack of support.

## Changelog

See our [CHANGELOG](./CHANGELOG.md).

## The Fine Print

Remember:

It is worth noting that this is **NOT SUPPORTED** by Sonatype, and is a contribution of ours to the open source
community (read: you!)

* Use this contribution at the risk tolerance that you have
* Do NOT file Sonatype support tickets related to `ossindex-lib`
* DO file issues here on GitHub, so that the community can pitch in

Phew, that was easier than I thought. Last but not least of all - have fun!


CHANGELOG.md
# Changelog

<!--next-version-placeholder-->

## v1.1.1 (2022-09-12)
### Fix
* Relax depenency on PyYAML ([`85e689b`](https://github.com/sonatype-nexus-community/ossindex-python/commit/85e689bdcb35df4b9e59b4ad1987242ccea94415))

## v1.1.0 (2022-07-12)
### Feature
* Support providing authentication directly over loading from configuration file ([`1bda4a9`](https://github.com/sonatype-nexus-community/ossindex-python/commit/1bda4a94e00fca30cf7488f1cb1e1bfaadaf1676))
* Support providing authentication directly over loading from configuration file ([`bcd86cb`](https://github.com/sonatype-nexus-community/ossindex-python/commit/bcd86cb70a6e889d5b34d9fdb2c58f77684f7b6d))

## v1.0.0 (2022-03-10)
### Feature
* Adopted PEP-561 #4 ([`f4b8b01`](https://github.com/sonatype-nexus-community/ossindex-python/commit/f4b8b0172fabfb55362450487d9cddaeaed3dc92))
* Added support for authentication to OSS Index #1 ([`aa26387`](https://github.com/sonatype-nexus-community/ossindex-python/commit/aa263872aeeb0ce3aa6c93de11107397f72cfb0b))
* Re-worked how we use TinyDB to attempt to resolve #2 ([`d5564da`](https://github.com/sonatype-nexus-community/ossindex-python/commit/d5564da8aaa207806e4a04db94c247ceec5fec2b))

### Fix
* Typo in `requirements.lowest.txt` ([`696f770`](https://github.com/sonatype-nexus-community/ossindex-python/commit/696f770052486fcc194a686f06ca5b47ff1f1a23))
* Potential fix for #2 ([`46f705d`](https://github.com/sonatype-nexus-community/ossindex-python/commit/46f705d117de799443902c2cb312f03b61a1ab44))
* Bug in deserialisation from cache database resolved ([`e3b9a9d`](https://github.com/sonatype-nexus-community/ossindex-python/commit/e3b9a9d74c541dead24aad90a0046fae078affca))
* Typing and imports ([`31a7e37`](https://github.com/sonatype-nexus-community/ossindex-python/commit/31a7e37e334301896f8275fb3e4f90119c05185b))

### Breaking
* #4 ([`f4b8b01`](https://github.com/sonatype-nexus-community/ossindex-python/commit/f4b8b0172fabfb55362450487d9cddaeaed3dc92))

## v0.2.1 (2021-09-16)
### Fix
* Typo in README ([`a260e91`](https://github.com/sonatype-nexus-community/ossindex-python/commit/a260e912227895d42b01a4775e36e7124fbc6f0b))

## v0.2.0 (2021-09-15)
### Feature
* Added accessor methods to Vulnerability class + helper method on OssIndexComponent to get the overriding maximum CVSS Score ([`15870a1`](https://github.com/sonatype-nexus-community/ossindex-python/commit/15870a19c7e62efb7f1c10f8f785f9d272152fc9))

### Fix
* Test corrected as was wrong ([`669ec24`](https://github.com/sonatype-nexus-community/ossindex-python/commit/669ec2481e6bbf46337e10149086c2b1fbba10e3))
* Corrected typing for OSS Index URL ([`813e593`](https://github.com/sonatype-nexus-community/ossindex-python/commit/813e593ef940cb27430590f80e2ad8628333fb2a))
* Removed garbage typo ([`23e7425`](https://github.com/sonatype-nexus-community/ossindex-python/commit/23e7425fddf104b332865f4382a9e4d857e76c04))
* Correct logic for determining which Vulnerability has the highest CVS score ([`37e5aed`](https://github.com/sonatype-nexus-community/ossindex-python/commit/37e5aed6424594f6efff136c7037d1e5db28ccd5))
* Added support for cwe being optional and added support for optional cve being present ([`3efafa9`](https://github.com/sonatype-nexus-community/ossindex-python/commit/3efafa978a2c9af647b760b34e4900917372a8d9))
* Support for OSS Index not returning descriptions for components ([`8244735`](https://github.com/sonatype-nexus-community/ossindex-python/commit/82447357188ee9c4f0a8c046b9d6a00ea556c7fc))

## v0.1.1 (2021-09-13)
### Fix
* Deployment GH action ([`1d403b5`](https://github.com/sonatype-nexus-community/ossindex-python/commit/1d403b565cc61ac100e7b8fc53a9f18c55e3c793))
* **doc:** Updated documentation to reflect usage and todos. ([`0078668`](https://github.com/sonatype-nexus-community/ossindex-python/commit/0078668a76f1911b349dccd0c21cbdcfb7dc5097))


tests/data.py
#
# Copyright 2022-Present Sonatype Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
from ossindex.model import OssIndexComponent, Vulnerability


def component_pip_18_no_vulnerabilities() -> OssIndexComponent:
    return OssIndexComponent(
        coordinates='pkg:pypi/pip@18.1',
        description='The PyPA recommended tool for installing Python packages.',
        reference='https://ossindex.sonatype.org/component/pkg:pypi/pip@18.1?utm_source=mozilla&utm_medium=integration'
                  '&utm_content=5.0'
    )


def component_pip_18_with_multiple_vulnerabilities() -> OssIndexComponent:
    oic = component_pip_18_no_vulnerabilities()
    oic.vulnerabilities.add(
        Vulnerability(
            id_='6dcf0940-d3c7-480f-84de-e90c3631ea81',
            display_name='CVE-2019-20916',
            title='[CVE-2019-20916] The pip package before 19.2 for Python allows Directory Traversal when a URL is '
                  '...',
            description='The pip package before 19.2 for Python allows Directory Traversal when a URL is given in an '
                        'install command, because a Content-Disposition header can have ../ in a filename, as '
                        'demonstrated by overwriting the /root/.ssh/authorized_keys file. This occurs in '
                        '_download_http_url in _internal/download.py.',
            cvss_score=7.5,
            cvss_vector='CVSS:3.0/AV:N/AC:L/PR:N/UI:N/S:U/C:N/I:H/A:N',
            cve='CVE-2019-20916',
            reference='https://ossindex.sonatype.org/vulnerability/6dcf0940-d3c7-480f-84de-e90c3631ea81?'
                      'component-type=pypi&component-name=pip&utm_source=mozilla&utm_medium=integration'
                      '&utm_content=5.0',
            external_references=[
                'https://nvd.nist.gov/vuln/detail/CVE-2019-20916'
            ]
        ),
    )
    oic.vulnerabilities.add(
        Vulnerability(
            id_='e4c955a3-2004-472e-920b-783fea46c3cd',
            display_name='OSSINDEX-783f-ea46-c3cd',
            title='CWE-22: Improper Limitation of a Pathname to a Restricted Directory (\'Path Traversal\')',
            description='The software uses external input to construct a pathname that is intended to identify a file '
                        'or directory that is located underneath a restricted parent directory, but the software does '
                        'not properly neutralize special elements within the pathname that can cause the pathname to '
                        'resolve to a location that is outside of the restricted directory.',
            cvss_score=3.6,
            cvss_vector='CVSS:3.0/AV:L/AC:L/PR:N/UI:R/S:C/C:L/I:N/A:N',
            cwe='CWE-22',
            reference='https://ossindex.sonatype.org/vulnerability/e4c955a3-2004-472e-920b-783fea46c3cd?'
                      'component-type=pypi&component-name=pip&utm_source=mozilla&utm_medium=integration'
                      '&utm_content=5.0',
            external_references=[
                'https://github.com/pypa/pip/issues/730'
            ]
        )
    )
    return oic


tests/mocks.py
#
# Copyright 2022-Present Sonatype Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
import json
from typing import Callable, Optional


class MockResponse:
    def __init__(self, data: Optional[str], status_code: int) -> None:
        self._text = data if data else ''
        self._status_code = status_code

    @property
    def status_code(self) -> int:
        return self._status_code

    @property
    def text(self) -> str:
        return self._text

    def json(self, object_hook: Callable) -> object:
        return json.loads(self.text, object_hook=object_hook)


def mock_oss_index_post(*args, **kwargs) -> MockResponse:
    if 'url' in kwargs.keys() and kwargs['url'] == 'https://ossindex.sonatype.org/api/v3/component-report':
        request_json = kwargs['json'] if 'json' in kwargs.keys() else {}

        if 'coordinates' in request_json.keys() and len(request_json['coordinates']) > 0:
            mock_response_data = []
            if 'pkg:pypi/pip@0.0.7' in request_json['coordinates']:
                return MockResponse(None, 401)

            if 'pkg:pypi/pip@21.2.3' in request_json['coordinates']:
                mock_response_data.append({
                    "coordinates": "pkg:pypi/pip@21.2.3",
                    "description": "The PyPA recommended tool for installing Python packages.",
                    "reference": "https://ossindex.sonatype.org/component/pkg:pypi/pip@21.2.3?"
                                 "utm_source=mozilla&utm_medium=integration&utm_content=5.0",
                    "vulnerabilities": []
                })

            if 'pkg:pypi/cryptography@3.3.1' in request_json['coordinates']:
                mock_response_data.append({
                    "coordinates": "pkg:pypi/cryptography@3.3.1",
                    "description": "cryptography is a package which provides cryptographic recipes and primitives to "
                                   "Python developers.",
                    "reference": "https://ossindex.sonatype.org/component/pkg:pypi/cryptography@3.3.1?"
                                 "utm_source=mozilla&utm_medium=integration&utm_content=5.0",
                    "vulnerabilities": [
                        {
                            "id": "333aca51-7375-4a9d-be64-16d316ab9274",
                            "displayName": "CVE-2020-36242",
                            "title": "[CVE-2020-36242] In the cryptography package before 3.3.2 for Python, certain "
                                     "sequences of update...",
                            "description": "In the cryptography package before 3.3.2 for Python, certain sequences of "
                                           "update calls to symmetrically encrypt multi-GB values could result in an "
                                           "integer overflow and buffer overflow, as demonstrated by the Fernet class.",
                            "cvssScore": 9.1,
                            "cvssVector": "CVSS:3.0/AV:N/AC:L/PR:N/UI:N/S:U/C:H/I:N/A:H",
                            "cve": "CVE-2020-36242",
                            "reference": "https://ossindex.sonatype.org/vulnerability/333aca51-7375-4a9d-be64-"
                                         "16d316ab9274?component-type=pypi&component-name=cryptography&"
                                         "utm_source=mozilla&utm_medium=integration&utm_content=5.0",
                            "externalReferences": [
                                "https://nvd.nist.gov/vuln/detail/CVE-2020-36242"
                            ]
                        }
                    ]
                })

            return MockResponse(json.dumps(mock_response_data), 200)

        return MockResponse(None, 404)


tests/__init__.py
#
# Copyright 2022-Present Sonatype Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
import os

FIXTURES_DIRECTORY = os.path.join(os.path.dirname(__file__), 'fixtures')


ossindex/model.py
#
# Copyright 2022-Present Sonatype Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
from typing import Iterable, Optional, Set

# See https://github.com/package-url/packageurl-python/issues/65
from packageurl import PackageURL  # type: ignore


class Vulnerability:
    """
    Model class that represents a Vulnerability as received back from OSS Index.

    """

    def __init__(self, *, id_: str, display_name: str, title: str, description: str, cvss_score: Optional[float] = None,
                 cvss_vector: Optional[str] = None, cve: Optional[str] = None, cwe: Optional[str] = None,
                 version_ranges: Optional[Iterable[str]] = None, reference: str,
                 external_references: Optional[Iterable[str]] = None):
        self.id = id_
        self.display_name = display_name
        self.title = title
        self.description = description
        self.cvss_score = cvss_score
        self.cvss_vector = cvss_vector
        self.cve = cve
        self.cwe = cwe
        self.version_ranges = set(version_ranges or [])
        self.reference = reference
        self.external_references = set(external_references or [])

    @property
    def id(self) -> str:
        """
        OSS Index's unique UUID for this Vulnerability.

        Returns:
             `str`
        """
        return self._id

    @id.setter
    def id(self, id_: str) -> None:
        self._id = id_

    @property
    def display_name(self) -> str:
        """
        displayName returned by OSS Index

        Returns:
            `str`
        """
        return self._display_name

    @display_name.setter
    def display_name(self, display_name: str) -> None:
        self._display_name = display_name

    @property
    def title(self) -> str:
        """
        title returned by OSS Index

        Returns:
            `str`
        """
        return self._title

    @title.setter
    def title(self, title: str) -> None:
        self._title = title

    @property
    def description(self) -> str:
        """
        description returned by OSS Index.

        Returns:
             `str`
        """
        return self._description

    @description.setter
    def description(self, description: str) -> None:
        self._description = description

    @property
    def cvss_score(self) -> Optional[float]:
        """
        CVSS Score returned from OSS Index.

        Returns:
             `float` if set else `None`
        """
        return self._cvss_score

    @cvss_score.setter
    def cvss_score(self, cvss_score: Optional[float]) -> None:
        self._cvss_score = cvss_score

    @property
    def cvss_vector(self) -> Optional[str]:
        """
        CVSS Vector returned from OSS Index

        Returns:
            `str` if set else `None`
        """
        return self._cvss_vector

    @cvss_vector.setter
    def cvss_vector(self, cvss_vector: Optional[str]) -> None:
        self._cvss_vector = cvss_vector

    @property
    def cwe(self) -> Optional[str]:
        """
        CWE returned from OSS Index.

        .. note:
            This is a string of the format CWE-nnn or an empty string

        Returns:
             `str` if set else `None`
        """
        return self._cwe

    @cwe.setter
    def cwe(self, cwe: Optional[str]) -> None:
        self._cwe = cwe

    @property
    def cve(self) -> Optional[str]:
        """
        CVE returned from OSS Index.

        Returns:
             `str` if set else `None`
        """
        return self._cve

    @cve.setter
    def cve(self, cve: Optional[str]) -> None:
        self._cve = cve

    @property
    def reference(self) -> str:
        """
        Reference URL to OSS Index for this Vulnerability.

        Returns:
            `str`
        """
        return self._reference

    @reference.setter
    def reference(self, reference: str) -> None:
        self._reference = reference

    @property
    def version_ranges(self) -> Set[str]:
        """
        Range of versions which are impacted by this Vulnerability.

        Returns:
            Set of `str`
        """
        return self._version_ranges

    @version_ranges.setter
    def version_ranges(self, version_ranges: Iterable[str]) -> None:
        self._version_ranges = set(version_ranges)

    @property
    def external_references(self) -> Set[str]:
        """
        List of external references that provide additional information about the vulnerability.

        Returns:
            Set of `str`
        """
        return self._external_references

    @external_references.setter
    def external_references(self, external_references: Iterable[str]) -> None:
        self._external_references = set(external_references)

    def __eq__(self, other: object) -> bool:
        if isinstance(other, Vulnerability):
            return hash(other) == hash(self)
        return False

    def __hash__(self) -> int:
        return hash((
            self.id, self.display_name, self.title, self.description, self.cvss_score, self.cvss_vector, self.cve,
            self.cwe, tuple(self.version_ranges), self.reference, tuple(self.external_references)
        ))

    def __repr__(self) -> str:
        return f'<Vulnerability id={self.id}, name={self.display_name}, cvss_score={self.cvss_score:.1f}>'


class OssIndexComponent:
    """
    Model class that represents a Component Report as received back from OSS Index.

    """

    def __init__(self, *, coordinates: str, description: Optional[str] = None, reference: str,
                 vulnerabilities: Optional[Iterable[Vulnerability]] = None) -> None:
        self.coordinates = coordinates
        self.description = description
        self.reference = reference
        self.vulnerabilities = set(vulnerabilities or [])

    @property
    def coordinates(self) -> str:
        """
        PackageURL formatted coordinates of this Component.

        Returns:
             `str`
        """
        return self._coordinates

    @coordinates.setter
    def coordinates(self, coordinates: str) -> None:
        self._coordinates = coordinates

    @property
    def description(self) -> Optional[str]:
        """
        Description of the Component from OSS Index.

        Returns:
            `str` if set else `None`
        """
        return self._description

    @description.setter
    def description(self, description: Optional[str]) -> None:
        self._description = description

    @property
    def reference(self) -> str:
        """
        URL to this Component on OSS Index.

        Returns:
             `str`
        """
        return self._reference

    @reference.setter
    def reference(self, reference: str) -> None:
        self._reference = reference

    @property
    def vulnerabilities(self) -> Set[Vulnerability]:
        """
        Known vulnerabilities that relate to this Component.

        Returns:
             Set of `Vulnerability`
        """
        return self._vulnerabilities

    @vulnerabilities.setter
    def vulnerabilities(self, vulnerabilities: Iterable[Vulnerability]) -> None:
        self._vulnerabilities = set(vulnerabilities)

    def __eq__(self, other: object) -> bool:
        if isinstance(other, OssIndexComponent):
            return hash(other) == hash(self)
        return False

    def __hash__(self) -> int:
        return hash((
            self.coordinates, self.description, self.reference, tuple(self.vulnerabilities)
        ))

    def __repr__(self) -> str:
        return f'<OssIndexComponent coordinates={self.coordinates}>'

    def get_package_url(self) -> PackageURL:
        """
        Get a PURL representation of this components coordinates.

        Returns:
            `PackageURL`
        """
        return PackageURL.from_string(purl=self.coordinates)

    def get_max_cvss_score(self) -> float:
        """
        Get the maximum CVSS Score across all Vulnerabilities known for this Component.

        Returns:
             `float`
        """
        max_cvss_score = 0.0
        if self.vulnerabilities:
            for v in self.vulnerabilities:
                max_cvss_score = OssIndexComponent._reduce_on_max_cvss_score(v=v, current_max=max_cvss_score)
        return max_cvss_score

    @staticmethod
    def _reduce_on_max_cvss_score(v: Vulnerability, current_max: float) -> float:
        if v.cvss_score:
            if v.cvss_score > current_max:
                return v.cvss_score
        return current_max


ossindex/ossindex.py
#
# Copyright 2022-Present Sonatype Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
import json
import logging
import os
import sys
from datetime import datetime, timedelta
from pathlib import Path
from typing import Dict, List, Optional, Tuple

import requests
import yaml
# See https://github.com/package-url/packageurl-python/issues/65
from packageurl import PackageURL  # type: ignore
from tinydb import TinyDB, Query
from tinydb.table import Document

from .exception import AccessDeniedException
from .model import OssIndexComponent
from .serializer import json_decoder, OssIndexJsonEncoder

logger = logging.getLogger('ossindex')

if sys.version_info >= (3, 8):
    from importlib.metadata import version as meta_version
else:
    from importlib_metadata import version as meta_version

ossindex_lib_version: str = 'TBC'
try:
    ossindex_lib_version = str(meta_version('ossindex-lib'))  # type: ignore[no-untyped-call]
except Exception:
    ossindex_lib_version = 'DEVELOPMENT'


class OssIndex:
    DEFAULT_CONFIG_FILE = '.oss-index.config'

    _caching_enabled: bool = False
    _cache_directory: str = '.ossindex'
    _cache_ttl_in_hours: int = 12

    _oss_index_api_version: str = 'v3'
    _oss_index_host: str = 'https://ossindex.sonatype.org'
    _oss_max_coordinates_per_request: int = 128
    _oss_index_authentication: Optional[requests.auth.HTTPBasicAuth] = None

    def __init__(self, *, enable_cache: bool = True, cache_location: Optional[str] = None,
                 username: Optional[str] = None, password: Optional[str] = None) -> None:
        self._caching_enabled = enable_cache
        if self._caching_enabled:
            logger.info('OssIndex caching is ENABLED')
            self._setup_cache(cache_location=cache_location)

        if username and password:
            logger.debug('Using supplied credentials for OSS Index')
            self._oss_index_authentication = requests.auth.HTTPBasicAuth(username, password)
        else:
            logger.debug('Attempting to load credentials for OSS Index from configuration file')
            self._attempt_config_load()

    def has_ossindex_authentication(self) -> bool:
        return self._oss_index_authentication is not None

    def _attempt_config_load(self) -> None:
        try:
            config_filename: str = os.path.join(os.path.expanduser('~'), OssIndex.DEFAULT_CONFIG_FILE)
            with open(config_filename, 'r') as ossindex_confg_f:
                ossindex_config = yaml.safe_load(ossindex_confg_f.read())
                self._oss_index_authentication = requests.auth.HTTPBasicAuth(
                    ossindex_config['username'], ossindex_config['password']
                )
        except FileNotFoundError:
            pass

    def get_component_report(self, packages: List[PackageURL]) -> List[OssIndexComponent]:
        logger.debug('A total of {} Packages to be queried against OSS Index'.format(len(packages)))
        return self._get_results(packages=packages)

    def purge_local_cache(self) -> None:
        if self._caching_enabled:
            logger.info('Truncating local cache database as requested')
            with self._get_cache_db() as db:
                db.truncate()
                logger.info('Local OSS Index cache has been purged')

    def _chunk_packages_for_oss_index(self, packages: List[PackageURL]) -> List[List[PackageURL]]:
        """
        Splits up the list of packages into lists that are of a size consumable by OSS Index
        APIs.

        :param packages: List[PackageURL]
        :return: List[List[PackageURL]]
        """
        return list(
            [packages[i: i + self._oss_max_coordinates_per_request] for i in
             range(0, len(packages), self._oss_max_coordinates_per_request)]
        )

    def _get_api_url(self, api_uri: str) -> str:
        return '{}/api/{}/{}'.format(
            self._oss_index_host, self._oss_index_api_version, api_uri
        )

    def _get_cached_results(self, packages: List[PackageURL]) -> Tuple[List[PackageURL], List[OssIndexComponent]]:
        """
        Takes a list of packages and returns two Lists:
            1. Packages without cached results
            2. Cached results for those packages where they exist

        :param packages: List[PackageURL]
        :return: (List[PackageURL], List[OssIndexComponent])
        """
        if not self._caching_enabled:
            # This should not be possible, but adding for developer safety
            return packages, []

        now = datetime.now()
        cached_results: List[OssIndexComponent] = []
        non_cached_packaged: List[PackageURL] = []
        with self._get_cache_db() as cache_db:
            for package in packages:
                logger.debug('   Checking in cache for {}'.format(package.to_string()))
                cache_results: List[Document] = cache_db.search(Query().coordinates == package.to_string())
                if len(cache_results) == 0:
                    # Not cached
                    logger.debug('      Not in cache')
                    non_cached_packaged.append(package)
                elif datetime.strptime(cache_results[0]['expiry'], '%Y-%m-%dT%H:%M:%S.%f') < now:
                    logger.debug('      Cached, but result expired')
                    non_cached_packaged.append(package)
                else:
                    logger.debug('      Cached, loading from cache')
                    cached_results.append(
                        json.loads(f'[{cache_results[0]["response"]}]', object_hook=json_decoder).pop()
                    )

        return non_cached_packaged, cached_results

    @staticmethod
    def _get_headers() -> Dict[str, str]:
        return {
            'Accept': 'application/vnd.ossindex.component-report.v1+json',
            'Content-type': 'application/vnd.ossindex.component-report-request.v1+json',
            'User-Agent': f'python-oss-index-lib@{ossindex_lib_version}'
        }

    def _get_results(self, packages: List[PackageURL]) -> List[OssIndexComponent]:
        results: List[OssIndexComponent] = list()

        # First get any cached results
        if self._caching_enabled:
            logger.debug('Checking local cache for any usable results...')
            packages, results = self._get_cached_results(packages=packages)
            logger.debug('   {} cached results found leaving {} to ask OSS Index'.format(
                len(results), len(packages)
            ))

        # Second, chunk up packages for which we have no cached results and query OSS Index
        chunk: List[PackageURL]
        chunks = self._chunk_packages_for_oss_index(packages=packages)
        logger.debug('Split {} packages into {} chunks for OSS requests'.format(len(packages), len(chunks)))
        for chunk in chunks:
            logger.debug('  Getting chunk results from OSS Index...')
            results = results + self._make_oss_index_component_report_call(packages=chunk)

        logger.debug('Total of {} results (including cached)'.format(len(results)))
        return results

    def _make_oss_index_component_report_call(self, packages: List[PackageURL]) -> List[OssIndexComponent]:
        response = requests.post(
            url=self._get_api_url('component-report'),
            headers=self._get_headers(),
            json={
                'coordinates': list(map(lambda p: str(p.to_string()), packages))
            },
            auth=self._oss_index_authentication
        )

        if not response.status_code == 200:
            raise AccessDeniedException()

        results: List[OssIndexComponent] = []
        for oic in response.json(object_hook=json_decoder):
            results.append(oic)

        if self._caching_enabled:
            self._upsert_cache_with_oss_index_responses(oss_components=results)
        return results

    def _upsert_cache_with_oss_index_responses(self, oss_components: List[OssIndexComponent]) -> None:
        if not self._caching_enabled:
            return

        now = datetime.now()
        cache_expiry = now + timedelta(hours=self._cache_ttl_in_hours)
        oc: OssIndexComponent
        with self._get_cache_db() as cache_db:
            for oc in oss_components:
                query: Query = Query()
                cache_query_result: List[Document] = cache_db.search(query.coordinates == oc.coordinates)
                if len(cache_query_result) == 0:
                    # New component for caching
                    logger.debug('    Caching new Component results for {}'.format(oc.coordinates))
                    cache_db.insert({
                        'coordinates': oc.coordinates,
                        'response': json.dumps(oc, cls=OssIndexJsonEncoder),
                        'expiry': cache_expiry.isoformat()
                    })
                else:
                    # Update existing cache
                    logger.debug('    Might refresh cache for {}'.format(oc.coordinates))
                    if now > datetime.strptime(cache_query_result[0]['expiry'], '%Y-%m-%dT%H:%M:%S.%f'):
                        # Cache expired - update it!
                        logger.debug('        Cache expired for {} - UPDATING CACHE'.format(oc.coordinates))
                        cache_db.update({
                            'response': json.dumps(oc, cls=OssIndexJsonEncoder),
                            'expiry': cache_expiry.isoformat()
                        }, query.coordinates == oc.coordinates)
                    else:
                        logger.debug('    Cache is still valid for {} - not updating'.format(oc.coordinates))

    def _get_cache_db(self) -> TinyDB:
        return TinyDB(os.path.join(self._cache_directory, 'ossindex.json'))

    def _setup_cache(self, cache_location: Optional[str] = None) -> None:
        full_cache_path: str
        if not cache_location:
            full_cache_path = os.path.join(Path.home(), self._cache_directory)
        else:
            full_cache_path = os.path.join(cache_location, self._cache_directory)

        if not os.path.exists(full_cache_path):
            Path(full_cache_path).mkdir(parents=True, exist_ok=True)

        self._cache_directory = str(Path(full_cache_path))
